fe/loss.py

```diff
@@ -21,7 +21,7 @@ def mybar_impl(w):
     return A
 
 
-def mybar_vjp(g, w):
+def mybar_jvp(g, w):
```

Comment:
In general, a vjp is not implemented in the same way as a jvp. In particular, a vjp is a vector jacobian product, and a jvp is a jacobian vector product. The former is a left multiply, as shown here, where as the latter is a right multiply.

Reply:
Furthermore, `g` is usually the adjoint, i.e. the derivative of the loss with respect to the output for the underlying function. Wher as in jvps, the gradient is the derivative of the input of the underlying function with respect to the parameters.

Reply:
Admittedly I was not very sure of what I was doing here.. From what I can gather the `defvjp` method we're using here was removed from jax ([commit](https://github.com/google/jax/issues/4918#issuecomment-728689031)). For untraceable code, I did find a [suggestion](https://github.com/google/jax/issues/4918#issuecomment-728689031) to define both jvp and a "transposition rule", but wasn't able to figure out the latter yet. However, I was surprised to find that `grad` still worked (with the same output and apparently performance) when I simply changed `jvp` to `vjp` :thinking: . I'll try to understand this better before moving this PR out of "draft" status.

Reply:
Consider a function `f(g(h(x)))`, where `f`,`g`,`h` are vector valued functions, and `x` is a vector - its derivative is `df/dg * dg/dh * dh/dx`and `*` denotes a matrix dot product. One way to evaluate this is:

`d/dx f = df/dg * (dg/dh * (dh/dx))`, i.e. inside out. This is "forward-mode", where `jv-product` here is comprised of: `j = dg/dh`, `v = dh/dx`, so `dg/dx = jv`
Then we apply the same `jvp` rule once more, this time setting `v` to the result of the above output:
`df/dg * dg/dx`, with `j=df/dg` and `v=dg/dx`. Here, the `v` is the intermediate that is carried along, or propagated forwards. In this case, `v` corresponds to the derivative of the intermediate with respect to the model parameters.

An alternative way is to evaluate it is outside in:

`d/dx f = ((df/dg) * dg/dh) * dh/dx`

Here we use reverse-mode, via `vj-products`, where `v=df/dg` and `j=dg/dh` in the first pass, and in the second pass `v=df/dh` and `j=dh/dx`. In this notation, the "adjoint" is refers to the `v` and is typically carried along (or back-propagated). In other words, in backwards mode, the `v` is the derivative of the final output "loss" with respect to the intermediate.

There are certain cases where the `vjp` and the `jvp` are identical. Eg. when the `j` is a symmetric matrix then any `v` would generate identical results. Or when `v` is a scalar. However, in your case, you simply misnamed the arguments.

`def_vjp` has the signature `foo(adjoint_grad, val)`, where as `def_jvp` has the signature `foo(primals, tangents)`, which is another way of saying `foo(val, grad)`. However, because the primitive `G(w): R^N - >R` emits a scalar valued output, its jacobian `dG/dw` is just a length N vector. It doesn't matter if it is "left multiplied" or "right multiplied" by a scalar.

Reply:
IIRC, the (offline) resolution of this was that `ad.defjvp` and `ad.defvjp` both take the gradient as the first argument (confusingly different from `custom_jvp` and `custom_vjp`, whose arguments are swapped with respect to each other). As a result the above change is (accidentally) correct.